File: src/main/dart/net/coderextreme/personal/geobubbles.dart
// Copyright 2021 by John Carlson.  All rights reserved.
import 'dart:io';
import '../x3d.dart';
var X3D0 =  X3D(

      profile_ : SFString('Immersive'),
      version_ : SFString('3.3'),
      head_ : 
        head(
          component_ : 
            component(
              name_ : SFString('Geospatial'),
              level_ : 1),
          meta_ : [
            meta(
              name_ : SFString('title'),
              content_ : SFString('geobubbles.x3d')),

            meta(
              name_ : SFString('creator'),
              content_ : SFString('John Carlson')),

            meta(
              name_ : SFString('generator'),
              content_ : SFString('manual')),

            meta(
              name_ : SFString('identifier'),
              content_ : SFString('https://coderextreme.net/X3DJSONLD/src/main/data/geobubbles.x3d')),

            meta(
              name_ : SFString('description'),
              content_ : SFString('geo bubbles'))]),
      Scene_ : 
        Scene(
          /*Viewpoint DEF='Tour' position='0 0 4' orientation='1 0 0 0' description='Tour Views'/*/
          /*PositionInterpolator DEF='TourPosition' key='0 1' keyValue='-0.5 -0.5 4 -0.5 0.5 4'/*/
          children_ : [
            GeoViewpoint(
              DEF_ : SFString('Tour'),
              position_ : SFVec3d([SFDouble(0), SFDouble(0), SFDouble(4)]),
              orientation_ : SFRotation([SFDouble(1), SFDouble(0), SFDouble(0), SFDouble(0)]),
              description_ : SFString('Tour Views')),

            Background(
              backUrl_ : MFString([SFString("../resources/images/BK.png"), SFString("https://coderextreme.net/X3DJSONLD/src/main/resources/images/BK.png")]),
              bottomUrl_ : MFString([SFString("../resources/images/BT.png"), SFString("https://coderextreme.net/X3DJSONLD/src/main/resources/images/BT.png")]),
              frontUrl_ : MFString([SFString("../resources/images/FR.png"), SFString("https://coderextreme.net/X3DJSONLD/src/main/resources/images/FR.png")]),
              leftUrl_ : MFString([SFString("../resources/images/LF.png"), SFString("https://coderextreme.net/X3DJSONLD/src/main/resources/images/LF.png")]),
              rightUrl_ : MFString([SFString("../resources/images/RT.png"), SFString("https://coderextreme.net/X3DJSONLD/src/main/resources/images/RT.png")]),
              topUrl_ : MFString([SFString("../resources/images/TP.png"), SFString("https://coderextreme.net/X3DJSONLD/src/main/resources/images/TP.png")])),

            Transform(
              child_ : 
                Shape(
                  geometry_ : 
                    Sphere(),
                  appearance_ : 
                    Appearance(
                      material_ : 
                        Material(
                          diffuseColor_ : SFColor([SFDouble(0.7), SFDouble(0.7), SFDouble(0.7)]),
                          specularColor_ : SFColor([SFDouble(0.5), SFDouble(0.5), SFDouble(0.5)]))))),

            TimeSensor(
              DEF_ : SFString('TourTime'),
              cycleInterval_ : 5,
              loop_ : true),

            GeoPositionInterpolator(
              DEF_ : SFString('TourPosition'),
              key_ : MFFloat([SFFloat(0), SFFloat(1)]),
              keyValue_ : MFVec3d([SFVec3d([0.0015708,0,4]),SFVec3d([0,0.0015708,4])])),

            Script(
              DEF_ : SFString('RandomTourTime'),
              field_ : [
                field(
                  type_ : SFString("SFTime"),
                  name_ : SFString('set_cycle'),
                  accessType_ : SFString("inputOnly")),

                field(
                  type_ : SFString("SFFloat"),
                  name_ : SFString('val'),
                  accessType_ : SFString("inputOutput"),
                  value_ : SFString('0')),

                field(
                  type_ : SFString("MFVec3d"),
                  name_ : SFString('positions'),
                  accessType_ : SFString("inputOutput"),
                  value_ : SFString('0.0015708 0 4 0 0.0015708 4')),

                field(
                  type_ : SFString("MFVec3d"),
                  name_ : SFString('position'),
                  accessType_ : SFString("inputOutput"),
                  value_ : SFString('0.0015708 0 4 0 0.0015708 4')),
              ],
ecmascript:eval (0

               , function set_cycle(value) {
                        var cartesianMult = -150;  // -150 if cartesian, 1 if geo
                        var ov = val;
			// Browser.print('old '+ov);
                        do {
                                val = Math.floor(Math.random()*2);
                                var vc = val;
                                positions[vc] = new SFVec3d(Math.round(Math.random()*2)*0.0015708*cartesianMult, Math.round(Math.random()*2)*0.0015708*cartesianMult, 4);
                        } while ( positions[ov][0] === positions[vc][0] && positions[ov][1] === positions[vc][1] && positions[ov][2] === positions[vc][2]);
			// Browser.println(positions[ov]);
			// Browser.println(positions[vc]);
                        position = new MFVec3d();
                        position[0] = new SFVec3d(positions[ov][0],positions[ov][1],positions[ov][2]);
                        position[1] = new SFVec3d(positions[vc][0],positions[vc][1],positions[vc][2]);
               })),

            ROUTE(
              fromNode_ : SFString('TourTime'),
              fromField_ : SFString('cycleTime'),
              toNode_ : SFString('RandomTourTime'),
              toField_ : SFString('set_cycle')),

            ROUTE(
              fromNode_ : SFString('RandomTourTime'),
              fromField_ : SFString('position'),
              toNode_ : SFString('TourPosition'),
              toField_ : SFString('keyValue')),

            ROUTE(
              fromNode_ : SFString('TourTime'),
              fromField_ : SFString('fraction_changed'),
              toNode_ : SFString('TourPosition'),
              toField_ : SFString('set_fraction')),

            ROUTE(
              fromNode_ : SFString('TourPosition'),
              fromField_ : SFString('geovalue_changed'),
              toNode_ : SFString('Tour'),
              toField_ : SFString('set_position'))]));
void main() { exit(0); }

File: src/main/python/net/coderextreme/personal/geobubbles.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "3.3"
head1 = x3d.head()
component2 = x3d.component()
component2.name = "Geospatial"
component2.level = 1

head1.children.append(component2)
meta3 = x3d.meta()
meta3.name = "title"
meta3.content = "geobubbles.x3d"

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "creator"
meta4.content = "John Carlson"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "generator"
meta5.content = "manual"

head1.children.append(meta5)
meta6 = x3d.meta()
meta6.name = "identifier"
meta6.content = "https://coderextreme.net/X3DJSONLD/src/main/personal/geobubbles.x3d"

head1.children.append(meta6)
meta7 = x3d.meta()
meta7.name = "description"
meta7.content = "geo bubbles"

head1.children.append(meta7)

X3D0.head = head1
Scene8 = x3d.Scene()
"""Viewpoint DEF='Tour' position='0 0 4' orientation='1 0 0 0' description='Tour Views'/"""
"""PositionInterpolator DEF='TourPosition' key='0 1' keyValue='-0.5 -0.5 4 -0.5 0.5 4'/"""
GeoViewpoint9 = x3d.GeoViewpoint()
GeoViewpoint9.DEF = "Tour"
GeoViewpoint9.position = [0,0,4]
GeoViewpoint9.orientation = [1,0,0,0]
GeoViewpoint9.description = "Tour Views"

Scene8.children.append(GeoViewpoint9)
Background10 = x3d.Background()
Background10.backUrl = ["../resources/images/BK.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/BK.png"]
Background10.bottomUrl = ["../resources/images/BT.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/BT.png"]
Background10.frontUrl = ["../resources/images/FR.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/FR.png"]
Background10.leftUrl = ["../resources/images/LF.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/LF.png"]
Background10.rightUrl = ["../resources/images/RT.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/RT.png"]
Background10.topUrl = ["../resources/images/TP.png","https://coderextreme.net/X3DJSONLD/src/main/resources/images/TP.png"]

Scene8.children.append(Background10)
Transform11 = x3d.Transform()
Shape12 = x3d.Shape()
Sphere13 = x3d.Sphere()

Shape12.geometry = Sphere13
Appearance14 = x3d.Appearance()
Material15 = x3d.Material()
Material15.diffuseColor = [0.7,0.7,0.7]
Material15.specularColor = [0.5,0.5,0.5]

Appearance14.material = Material15

Shape12.appearance = Appearance14

Transform11.children.append(Shape12)

Scene8.children.append(Transform11)
TimeSensor16 = x3d.TimeSensor()
TimeSensor16.DEF = "TourTime"
TimeSensor16.cycleInterval = 5
TimeSensor16.loop = True

Scene8.children.append(TimeSensor16)
GeoPositionInterpolator17 = x3d.GeoPositionInterpolator()
GeoPositionInterpolator17.DEF = "TourPosition"
GeoPositionInterpolator17.key = [0,1]

Scene8.children.append(GeoPositionInterpolator17)
Script18 = x3d.Script()
Script18.DEF = "RandomTourTime"
field19 = x3d.field()
field19.name = "set_cycle"
field19.accessType = "inputOnly"
field19.type = "SFTime"

Script18.field.append(field19)
field20 = x3d.field()
field20.name = "val"
field20.accessType = "inputOutput"
field20.type = "SFFloat"
field20.value = 0

Script18.field.append(field20)
field21 = x3d.field()
field21.name = "positions"
field21.accessType = "inputOutput"
field21.type = "MFVec3d"

Script18.field.append(field21)
field22 = x3d.field()
field22.name = "position"
field22.accessType = "inputOutput"
field22.type = "MFVec3d"

Script18.field.append(field22)

Script18.sourceCode = '''ecmascript:\n"+
"\n"+
"               function set_cycle(value) {\n"+
"                        var cartesianMult = -150;  // -150 if cartesian, 1 if geo\n"+
"                        var ov = val;\n"+
"			// Browser.print('old '+ov);\n"+
"                        do {\n"+
"                                val = Math.floor(Math.random()*2);\n"+
"                                var vc = val;\n"+
"                                positions[vc] = new SFVec3d(Math.round(Math.random()*2)*0.0015708*cartesianMult, Math.round(Math.random()*2)*0.0015708*cartesianMult, 4);\n"+
"                        } while ( positions[ov][0] === positions[vc][0] && positions[ov][1] === positions[vc][1] && positions[ov][2] === positions[vc][2]);\n"+
"			// Browser.println(positions[ov]);\n"+
"			// Browser.println(positions[vc]);\n"+
"                        position = new MFVec3d();\n"+
"                        position[0] = new SFVec3d(positions[ov][0],positions[ov][1],positions[ov][2]);\n"+
"                        position[1] = new SFVec3d(positions[vc][0],positions[vc][1],positions[vc][2]);\n"+
"               }'''

Scene8.children.append(Script18)
ROUTE23 = x3d.ROUTE()
ROUTE23.fromNode = "TourTime"
ROUTE23.fromField = "cycleTime"
ROUTE23.toNode = "RandomTourTime"
ROUTE23.toField = "set_cycle"

Scene8.children.append(ROUTE23)
ROUTE24 = x3d.ROUTE()
ROUTE24.fromNode = "RandomTourTime"
ROUTE24.fromField = "position"
ROUTE24.toNode = "TourPosition"
ROUTE24.toField = "keyValue"

Scene8.children.append(ROUTE24)
ROUTE25 = x3d.ROUTE()
ROUTE25.fromNode = "TourTime"
ROUTE25.fromField = "fraction_changed"
ROUTE25.toNode = "TourPosition"
ROUTE25.toField = "set_fraction"

Scene8.children.append(ROUTE25)
ROUTE26 = x3d.ROUTE()
ROUTE26.fromNode = "TourPosition"
ROUTE26.fromField = "geovalue_changed"
ROUTE26.toNode = "Tour"
ROUTE26.toField = "set_position"

Scene8.children.append(ROUTE26)

X3D0.Scene = Scene8
f = open("../personal/geobubbles.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../personal/geobubbles.new.python.json", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()
